### PR TITLE
Update docker container, add Docker auto-build and test, update documentation for building with spack-stack

### LINKED ISSUE #409: Update procedure for building SCM for spack-stack, clean up existing build procedures
- [x] Convert shell scripts to modules (if possible)
   - [x] Derecho, Hera: #406 
   - [x] Linux, MacOS: #466 
 - [x] Update Users Guide for the new build procedures #466 
   - Point to relevant documentation for spack stack
   - Update/rethink example submission scripts
 - ~~[ ] Update online tutorial~~ will be addressed by #456

### LINKED ISSUE #455: Docker container update
From @ligiabernardet 
> Mike will take a look.
> Not sure where/how the container is built and stored, but should be able to get access.

For past releases, Grant built docker image on windows machine and images stored on dockerhub.
Is DTC still hosting use of dockerhub?
Grant pass any info to Mike for reproducing.

Mike looked at current docker build - it won't even build as it is outdated
Sam suggests building docker with debian.

### PR DESCRIPTION
**Pull request by mkavulich**
This PR has several updates that improve SCM documentation, overhaul the docker container, updates new and existing tests, and adds modulefiles for building on MacOS and generic Linux.

### Documentation
Documentation for this branch is built here: https://ccpp-scm-mkavulich.readthedocs.io/en/feature-macos_and_docker_build/

This update adds some information about testing to the Repository chapter (though more details should probably be included), and completely overhauls the Quick Start chapter.

Among the changes to the Quick Start chapter are:

 - Removing some less relevant text, and rearranging and updating some wording to improve flow of the chapter (this is supposed to be a "quick" start after all)
 - Replace non-standard section delimiters per Sphinx documentation
 - Include more information about spack-stack, including more links
 - Include instructions for loading environment on pre-configured platforms, as well as other procedures for building spack-stack or setting environment variables manually on other platforms such as MacOS and Linux
 - Update Python section for appropriate prerequisites
 - Use Sphinx built-in Notes markup to highlight important points
 - Update run commands, add more examples
 - Update Docker section with relevant changes (see next section)

### Docker

 - The Dockerfile is updated to work with the latest CCPP code. The old "common community container" is no longer used, as it is no longer supported.
 - A github action is introduce that runs on each commit to `main`. This action will build the container and push it to Dockerhub under the tag "dtcenter/ccpp-scm:latest". This will ensure there is always an up-to-date, working container for the top of the main branch.

### Testing

 - A new CI test will build and run the regressiont tests with the Docker container when a new pull request is created
 - Test `ci_build_scm_ubuntu_22.04_nvidia.yml` is modified to only run on pull_requests and when manually requested

### Other

 - Add new "generic platform" modulefiles `macos_clang` and `linux_gnu` for use with spack-stack installations.

This PR will resolve #409 and partially resolve #455 (will still need a release container when the time comes)

### CODE REVIEW

**Comment by mkavulich** [comment]
Note: the new job to auto-push the latest container to Dockerhub makes use of repository secrets: I have provided my dockerhub credentials as DOCKERHUB_USERNAME and DOCKERHUB_TOKEN since I can write to the dtcenter organization on Dockerhub. If in the future I am not around, or if I lose access to that Dockerhub org, those secrets will have to be updated.

File: scm/doc/TechGuide/chap_quick.rst
Comment (dustinswales) [inline scm/doc/TechGuide/chap_quick.rst:126]: FYI. @scrasmussen plans to have these dependencies removed from the CCPP before the release.

File: scm/doc/TechGuide/chap_quick.rst
Comment (dustinswales) [inline scm/doc/TechGuide/chap_quick.rst:435]: With #465 we need to add
 `./contrib/get_aerosol_climo.sh`

File: scm/doc/TechGuide/chap_quick.rst
Comment (dustinswales) [inline scm/doc/TechGuide/chap_quick.rst:470]: Something we need to discuss are the SCM RTs.
[rt_test_cases.py](https://github.com/NCAR/ccpp-scm/blob/main/test/rt_test_cases.py) needs to be extended to 
- include the v7 tests, which we need to decide on. Same SDF/case combos as in v6 plus some new cases? 
(*Note* we have a separate CI test that use[ the DEPHY repository](https://github.com/GdR-DEPHY/DEPHY-SCM), but there aren't any comparisons in that test, it just ensures the SCM works with the DEPHY repo. When I set this up I didn't do a deep evaluation on the case from the DEPHY repo, just more of a technical test that teh SCM could use the latest and greatest DEPHY cases. @bluefinweiwei has used the SCM with cases in the DEPHY repo for T&E, so I think it makes more sense to use a case that she is confident in. And then add a comparison step to the DEPHY CI test.
(**Note** We have a separate CI test for the UFS case generation. This test just creates the UFS case files from staged UFS RT data. It doesn't run the SCM with that case. We need to either expand this test to run the SCM and compare to baselines, OR leave this CI test as is and add a staged UFS case to the SCM, to be included in rt_test_cases.py?)
- move the "supported v6" into the "unsupported" section. As long as these still work, we should include them in our testing suite.
- Beef up [the comparison step in the CI RT script](https://github.com/NCAR/ccpp-scm/blob/main/.github/workflows/ci_run_scm_rts.yml#L197) to create plots of fields that differ, and upload these as a github artifact.
- We only run the CI tests with GNU. Intel compilers aren't available on the guthub servers we do the CI, but we could create a Intel Docker container. TBH, for the RT CI test, we don't need to build the software stack as part of the test since [ci_build_scm_ubuntu_22.04.yaml](https://github.com/NCAR/ccpp-scm/blob/main/.github/workflows/ci_build_scm_ubuntu_22.04.yml) already does this, we just need to run the SCM and compare to baselines.

File: scm/doc/TechGuide/chap_intro.rst
Comment (dustinswales) [inline scm/doc/TechGuide/chap_intro.rst:106]: FWIW. We are going to need to change all instances of "replay" in the documentation to "case generation".
There will be a subsequent PR with those code changes, so no need to address this at the moment.

File: docker/Dockerfile
Comment (dustinswales) [inline docker/Dockerfile:1]: Should we name this file Dockerfile_GNU or Dockerfile_GNUv12? Just in case we get an Intel one soon.

**Review by dustinswales — APPROVED** [review]
Thanks for making these changes.
I just wonder if we should rip out the NCEPlibs stuff now from the documentation, or wait until that PR gets into the physics?

File: scm/doc/TechGuide/chap_quick.rst
Comment (mkavulich) [reply scm/doc/TechGuide/chap_quick.rst:435]: Should this be a separate script or does it make sense to roll it into the "get_all_static_data" script?

File: scm/doc/TechGuide/chap_quick.rst
Comment (mkavulich) [reply scm/doc/TechGuide/chap_quick.rst:79]: Fair point, I didn't think about this sounding negative when I wrote it but I can update the wording.

File: scm/doc/TechGuide/chap_intro.rst
Comment (mkavulich) [reply scm/doc/TechGuide/chap_intro.rst:106]: This link is currently broken (not caught in the last PR) so this just updates the link to point to the correct place.

File: docker/Dockerfile
Comment (mkavulich) [reply docker/Dockerfile:1]: Good point; do you think there is a possibility that Intel might be doable? Last time I worked with Docker (a few years ago) Intel was very difficult to get working in containers.

Comment (climbfuji) [reply docker/Dockerfile:1]: JCSDA uses Intel in Docker all the time. The only issue is that you can't distribute the containers to the public.

Comment (mkavulich) [reply docker/Dockerfile:1]: Ahh, okay that's good to know. If we do implement containers for testing we'll need a place to push them other than Dockerhub anyway, so it sounds like this may work.

**Comment by mkavulich** [comment]
@grantfirl Did you want a chance to review this or can I go ahead and merge it?

File: .github/workflows/build_and_push_docker_latest.yml
Comment (dustinswales) [inline .github/workflows/build_and_push_docker_latest.yml:14]: Maybe change to "ubuntu-22.04", the latest OS can change and cause hiccups.

File: scm/doc/TechGuide/chap_quick.rst
Comment (dustinswales) [inline scm/doc/TechGuide/chap_quick.rst:1]: We will need to revisit the documentation later anyways, at which point we can remove any deprecated dependencies.

**Review by dustinswales — APPROVED** [review]
Looks good to me. A minor suggestion with the OS version in the CI scripts. And we will need to revisit documentation later anyways.

File: .github/workflows/build_and_push_docker_latest.yml
Comment (mkavulich) [reply .github/workflows/build_and_push_docker_latest.yml]: Good point, updated

File: docker/Dockerfile
Comment (grantfirl) [inline docker/Dockerfile:47]: Why are we not trying to use spack-stack?

File: docker/Dockerfile
Comment (grantfirl) [inline docker/Dockerfile:77]: It's possible that we'll also need to add the ./get_aerosol_climo.sh script here too if we need to use GOCART climatology for anything in the Docker container. If the use case is just to show that it works, then I guess not, but if we need to run the HR3 suite that uses GOCART climatology, then we'll need to add it. Unfortunately, it's over 1 GB.

File: docker/Dockerfile
Comment (mkavulich) [reply docker/Dockerfile:47]: Spack-stack is a bit bloated for a container since it has a lot of software we don't use, especially since @scrasmussen is removing these NCEPLIBS dependencies soon.

File: scm/doc/TechGuide/chap_quick.rst
Comment (grantfirl) [inline scm/doc/TechGuide/chap_quick.rst:7]: Shortened because this is a "quick" start guide? Otherwise, this might be new information from someone unfamiliar with CCPP, etc.

File: docker/Dockerfile
Comment (mkavulich) [reply docker/Dockerfile:77]: I had thought about that...I think it's outside the scope of this PR, but one solution would be to "bind mount" a local directory to allow users to download this data separately outside the container (and not require downloading the data at build-time).

File: scm/doc/TechGuide/chap_quick.rst
Comment (grantfirl) [reply scm/doc/TechGuide/chap_quick.rst:49]: NVM. I see that it was just moved. Scrolling is hard. :)

File: scm/doc/TechGuide/chap_quick.rst
Comment (grantfirl) [inline scm/doc/TechGuide/chap_quick.rst:244]: Ah, I see that you reference the "manual" approach used in the Dockerfile. That's fine. Disregard the comment above re: spack-stack.

File: scm/doc/TechGuide/chap_quick.rst
Comment (grantfirl) [reply scm/doc/TechGuide/chap_quick.rst:435]: This should be separate, IMO, since it is larger than the others and it is not needed unless using GOCART climatology, which is only necessary for some configurations (e.g. latest HR3)

File: scm/doc/TechGuide/chap_quick.rst
Comment (grantfirl) [inline scm/doc/TechGuide/chap_quick.rst:444]: Yep.

File: scm/doc/TechGuide/chap_quick.rst
Comment (grantfirl) [inline scm/doc/TechGuide/chap_quick.rst:879]: The rt_test_cases.py has more than just the permutations of supported cases/suites, but that is probably not a big deal.

**Review by grantfirl — APPROVED** [review]

**Comment by grantfirl** [comment]
@mkavulich Can you update to the latest main for good measure before I press the merge button?

File: scm/doc/TechGuide/chap_quick.rst
Comment (mkavulich) [reply scm/doc/TechGuide/chap_quick.rst:7]: Yes that's the idea, I wanted to trim down the less relevant bits in this chapter to make it a "quicker" start. The "Intro" chapter already covers this at a high level, I could expand that intro with this info about submodules specifically if you think that's needed.

**Comment by mkavulich** [comment]
@grantfirl Rebased my branch on the latest main

**Pull request merged**
